### PR TITLE
Player and inventory

### PR DESCRIPTION
**Pull request by adam-kucz**
It sort of happened, same way as it did with GUI earlier.

I didn't have any idea what to do about ItemStorage though, it's arguably even slightly worse now (but I think it'll be easier to refactor it into something sensible from this form).

### CODE REVIEW

File: Scripts/GUI/GUIObjects/GUIInventorySlot.cs [hunk 1]
```diff
@@ -47,11 +47,11 @@ public GUIInventorySlot(GUIInventorySlot toExchangeOnPress, Item.ItemType t, int
...
     {
-        ItemStack floatingStack = exchangeSlot.GetCurItemStack();
-        ItemStack slotStack = this.GetCurItemStack();
+        ItemStack exchangeStack = exchangeSlot.GetCurItemStack();
+        ItemStack slotStack = GetCurItemStack();
 
-        if (floatingStack != null &&
-            !Item.CompatibleWith(floatingStack.Item.IType, this.GetItemType()))
+        if (exchangeStack != null &&
+            !Item.CompatibleWith(exchangeStack.Item.IType, this.GetItemType()))
```
Comment (JamesHinshelwood) [inline Scripts/GUI/GUIObjects/GUIInventorySlot.cs]: This could just be `!Item.CompatibleWith(exchangeStack?.Item.IType, this.GetItemType())` I think.

Comment (JamesHinshelwood) [reply Scripts/GUI/GUIObjects/GUIInventorySlot.cs]: Maybe not actually, that might cause the first term to return null, but the ItemType enum can't be null, hmmm

File: Scripts/GUI/GUIObjects/GUIInventorySlot.cs [hunk 2]
```diff
@@ -62,9 +62,17 @@ public override void OnLeftPress()
...
+        // TODO: think of a better conditional
+        if (exchangeStack != null && slotStack != null && slotStack.Item.Stackable && slotStack.Item.Id == exchangeStack.Item.Id)
+        {
+            slotStack.ChangeQuantity(exchangeStack.Count);
+            exchangeSlot.ClearItemStack();
+        }
+        else
+        {
+            exchangeSlot.AssignItemStack(slotStack);
+            AssignItemStack(exchangeStack);
```
Comment (JamesHinshelwood) [inline Scripts/GUI/GUIObjects/GUIInventorySlot.cs]: `this.AssignItemStack` might make this clearer

File: Scripts/GUI/GUIObjects/GUIInventorySlot.cs [hunk 3]
```diff
@@ -62,9 +62,17 @@ public override void OnLeftPress()
             return;
         }
 
-        // slotStack.Item.Stackable
-        exchangeSlot.AssignItemStack(slotStack);
-        this.AssignItemStack(floatingStack);
+        // TODO: think of a better conditional
```
Comment (JamesHinshelwood) [inline Scripts/GUI/GUIObjects/GUIInventorySlot.cs]: Can't think of anything better :man_shrugging:

**Review by JamesHinshelwood — CHANGES_REQUESTED** [review]
Player needs bars don't seem to be decreasing

The rest of the changes I commented aren't super important, so don't worry if you disagree

**Comment by adam-kucz** [comment]
I've added "this" back, it just got deleted with all the other "this"s that were actually redundant. Stats should now be decreasing, they were changing, it was that they were just increasing because I forgot the negative sign ;). As a bonus shift clicking now works with hand slot to inventory, will soon work everywhere.

File: Scripts/GUI/GUIObjects/GUIInventorySlot.cs (hunk 1 above)
Comment (adam-kucz) [reply Scripts/GUI/GUIObjects/GUIInventorySlot.cs]: It doesn't work like that, unfortunately 😞 

C# can't propagate null values (or as I like to call them the Maybe monad) through function calls (yet, hopefully) and that's what would be necessary here, I think.

**Review by JamesHinshelwood — APPROVED** [review]

**Comment by adam-kucz** [comment]
Could you take another look at the shift click now?

**Comment by JamesHinshelwood** [comment]
The item label is not removed when you shift click out of your hand (it still says CO2 Bacteria when I hover my mouse over the slot)

image

**Review by JamesHinshelwood — APPROVED** [review]

**Pull request merged**